hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs

```diff
@@ -198,22 +199,46 @@ instance ToJSON IdePlugins where
 
 -- ---------------------------------------------------------------------
 
-type IdeGhcM = GM.GhcModT IdeM
+type IdeGhcM = GM.GhcModT IdeBase
 
 instance MonadMTState IdeState IdeGhcM where
   readMTS = lift $ lift $ lift readMTS
-  modifyMTS f = lift $ lift $ lift $ modifyMTS f
+  modifyMTS = lift . lift . lift . modifyMTS
+
+-- | A computation that is deferred until the module is cached.
+-- Note that the module may not typecheck, in which case 'UriCacheFailed' is passed
+data IdeDefer a = IdeDefer FilePath (UriCache -> a) deriving Functor
```

Comment:
This is probably not the right time to do it, but at some point we should let the module cache be more lenient, in the sense that if we get anything out, we call the function.

Here anything is `ParsedSource`, `RenamedSource`, `TypeCheckedSource`.

Reply:
I wrote this review 5 days ago, but didn't realise you had to submit it for you to see it...